module/mempool/stdmap/incorporated_result_seals.go

```diff
@@ -9,18 +11,124 @@ import (
 // of the consensus nodes, used to store seals that need to be added to blocks.
 type IncorporatedResultSeals struct {
 	*Backend
+	// index the seals by the height of the executed block
+	byHeight     map[uint64][]flow.Identifier
+	lowestHeight uint64
+}
+
+func indexByHeight(seal *flow.IncorporatedResultSeal) uint64 {
+	return seal.Header.Height
 }
 
 // NewIncorporatedResults creates a mempool for the incorporated result seals
 func NewIncorporatedResultSeals(opts ...OptionFunc) *IncorporatedResultSeals {
-	return &IncorporatedResultSeals{
-		Backend: NewBackend(opts...),
+	r := &IncorporatedResultSeals{
+		Backend:  NewBackend(opts...),
+		byHeight: make(map[uint64][]flow.Identifier),
+	}
+
+	// assuming all the entities are for unsealed blocks, then we will remove a seal
+	// with the highest height.
+	r.eject = func(entities map[flow.Identifier]flow.Entity) (flow.Identifier, flow.Entity) {
+		maxHeight := uint64(0)
+		for height := range r.byHeight {
+			if height > maxHeight {
+				maxHeight = height
+			}
+		}
+		seals, ok := r.byHeight[maxHeight]
+		if !ok {
+			panic(fmt.Sprintf("can't find seals by height: %v", maxHeight))
+		}
+
+		if len(seals) == 0 {
+			panic(fmt.Sprintf("empty seal at height: %v", maxHeight))
+		}
+
+		sealID := seals[0]
+		seal, ok := entities[sealID]
+		if !ok {
+			panic(fmt.Sprintf("inconsistent index, can not find seal by id: %v", sealID))
+		}
+		return sealID, seal
+	}
+
+	// when eject a entity, also update the secondary indx
+	r.RegisterEjectionCallbacks(func(entity flow.Entity) {
+		seal := entity.(*flow.IncorporatedResultSeal)
+		removeSeal(seal, r.entities, r.byHeight)
+	})
+
+	return r
+}
+
+func removeSeal(seal *flow.IncorporatedResultSeal,
+	entities map[flow.Identifier]flow.Entity,
+	byHeight map[uint64][]flow.Identifier) {
+	sealID := seal.ID()
+	delete(entities, sealID)
+	index := indexByHeight(seal)
+	siblings := byHeight[index]
+	newsiblings := make([]flow.Identifier, 0, len(siblings))
+	for _, sibling := range siblings {
+		if sibling != sealID {
+			newsiblings = append(newsiblings, sibling)
+		}
+	}
+
+	if len(newsiblings) == 0 {
+		delete(byHeight, index)
+	} else {
+		byHeight[index] = newsiblings
 	}
 }
 
-// Add adds an IncorporatedResultSeal to the mempool
+func removeByHeight(height uint64, entities map[flow.Identifier]flow.Entity,
+	byHeight map[uint64][]flow.Identifier) error {
+	sameHeight, ok := byHeight[height]
+	if !ok {
+		return fmt.Errorf("cannot find seals by height: %v", height)
+	}
```

Comment:
This will almost certainly crash a consensus node: You are assuming that for every pruned height, there are actually seals in the mempool. There is no logic to guarantee this and this assumption is likely false fairly frequently.